api/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecution.java

```diff
@@ -158,7 +158,25 @@ public String getUrl() throws IOException {
      */
     public abstract void interrupt(Result r, CauseOfInterruption... causes) throws IOException, InterruptedException;
 
-    public abstract void addListener(GraphListener listener);
+    @Deprecated
+    public void addListener(GraphListener listener) {
+        addListener(listener, false);
+    }
+
+    /**
+     * Add a listener to changes in the flow graph structure.
+     * @param listener a listener to add
+     * @param synchronous true to receive events immediately as they occur (you must be very careful not to acquire locks or block); false to receive notification as soon as possible
+     */
+    public /*abstract*/ void addListener(GraphListener listener, boolean synchronous) {
```

Comment:
`synchronous` seems like something that should be configured on the `GraphListener` impl.

Yes, I see it's an `interface` => backward compatibility issues etc., but setting is as an arg on `addListener` smells bad to me.